rfcs/0003-radicle-daemon.rst

```diff
@@ -0,0 +1,98 @@
+Radicle daemon
+===============
+
+.. date:: 2019-01-09
+
+Synopsis
+---------
+
+This RFC gives a rough spec of the *radicle daemon*, a long-running background
+process which replicates remote machines on the users PC.
+
+Motivation
+-----------
+
+Currently to run a query against the state of some remote chain, all the inputs
+are downloaded and the state re-materialised from scratch. This is too slow.
+
+The daemon also allows apps not written in radicle or Haskell to have access to
+a radicle interpreter.
+
+Proposal
+---------
+
+The radicle daemon is an executable which starts a server which listens for HTTP
+requests on a specific port. The API uses JSON.
+
+- Query a machine, that is, run an expression against it.
+
+  ``POST /:machine_url/query``
+
+  .. code-block:: json
+
+    {"expression": "(get-state)",
+     "keep_input": 259200,
+     "keep_polling": 3600,
+     "poll_period": 30}
```

Comment:
Should all keys other than "expression" be optional? Otherwise I guess new queries would override old ones even if the new query doesn't care? Eg.

```
App 1:
POST /foo/query
{"expression": "(get-state)", "keep_polling": 3600}

App 2:
POST /foo/query
{"expression": "(get-state)", "keep_polling": 0} // Ok, no longer polling!

App 1:
POST /foo/query
{"expression": "(get-state)", "keep_polling": 3600} // Woops, this will be slow because the state is stale!

Reply:
Another way around this would be for apps or queries to have a unique token:
```
POST /foo/query?token=app1
{"expression": "(get-state)", "keep_polling": 3600}

POST /foo/query?token=app2
{"expression": "(get-state)", "keep_polling": 0}

POST /foo/query?token=app1
{"expression": "(get-state)", "keep_polling": 3600} // Fresh state because we didn't stop polling for app1!
```

Reply:
I like the token option.

Reply:
I also like the token option. To keep the RFC small we should add it to the “Open Questions” section.